include/vcpkg/base/parallel-algorithms.h

```diff
@@ -1,85 +1,163 @@
 
-    template<class Container, class F>
-    void parallel_for_each(Container&& c, F cb) noexcept
+        explicit operator bool() { return ptp_work != nullptr; }
+
+        void submit() { ::SubmitThreadpoolWork(ptp_work); }
+
+    private:
+        PTP_WORK ptp_work;
+    };
+
+    template<class F>
+    inline void execute_in_parallel(size_t work_count, F work) noexcept
     {
-        if (c.size() == 0)
+        if (work_count == 0)
         {
             return;
         }
-        if (c.size() == 1)
+
+        if (work_count == 1)
         {
-            cb(c[0]);
-            return;
+            work(size_t{});
         }
 
-        std::atomic_size_t next{0};
-
-        execute_in_parallel(c.size(), [&]() {
-            size_t i = 0;
-            while (i < c.size())
+        WorkCallbackContext<F> context{work, work_count};
+        PtpWork ptp_work([](PTP_CALLBACK_INSTANCE,
+                            void* context,
+                            PTP_WORK) noexcept { static_cast<WorkCallbackContext<F>*>(context)->run(); },
+                         &context,
+                         nullptr);
+        if (ptp_work)
+        {
+            auto max_threads = (std::min)(work_count, static_cast<size_t>(get_concurrency()));
+            // start at 1 to account for the running thread
+            for (size_t i = 1; i < max_threads; ++i)
             {
-                if (next.compare_exchange_weak(i, i + 1, std::memory_order_relaxed))
-                {
-                    cb(c[i]);
-                }
+                ptp_work.submit();
             }
-        });
+        }
+
+        context.run();
     }
+#else  // ^^^ _WIN32 / !_WIN32 vvv
+    struct JThread
+    {
+        template<class Arg0, std::enable_if_t<!std::is_same<JThread, std::decay_t<Arg0>>::value, int> = 0>
+        JThread(Arg0&& arg0) : m_thread(std::forward<Arg0>(arg0))
+        {
+        }
 
-    template<class Container, class RanItTarget, class F>
-    void parallel_transform(const Container& c, RanItTarget out_begin, F&& cb) noexcept
+        ~JThread() { m_thread.join(); }
+
+        JThread(const JThread&) = delete;
+        JThread& operator=(const JThread&) = delete;
+
+    private:
+        std::thread m_thread;
+    };
+
+    template<class F>
+    inline void execute_in_parallel(size_t work_count, F work) noexcept
     {
-        if (c.size() == 0)
+        if (work_count == 0)
         {
             return;
         }
-        if (c.size() == 1)
+
+        if (work_count == 1)
         {
-            *out_begin = cb(c[0]);
+            work(size_t{});
             return;
         }
 
-        std::atomic_size_t next{0};
-
-        execute_in_parallel(c.size(), [&]() {
-            size_t i = 0;
-            while (i < c.size())
+        WorkCallbackContext<F> context{work, work_count};
+        const size_t bg_threads = std::min(work_count, static_cast<size_t>(get_concurrency())) - 1;
+        std::vector<Optional<JThread>> workers(bg_threads);
```

Comment:
Why use Optional/resize/emplace vs reserve/emplace_back? Exception safety?

Reply:
emplace_back requires that T be movable, thread is not.

Reply:
> emplace_back requires that T be movable, thread is not.

`std::thread` is movable: https://en.cppreference.com/w/cpp/thread/thread/thread